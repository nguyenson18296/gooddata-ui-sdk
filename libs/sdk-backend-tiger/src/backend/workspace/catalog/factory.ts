// (C) 2019-2021 GoodData Corporation
import {
    IWorkspaceCatalog,
    IWorkspaceCatalogFactory,
    IWorkspaceCatalogFactoryOptions,
    CatalogItem,
    CatalogItemType,
    ICatalogFact,
    ICatalogGroup,
    ICatalogMeasure,
    IGroupableCatalogItemBase,
    isCatalogAttribute,
    isCatalogFact,
    isCatalogMeasure,
} from "@gooddata/sdk-backend-spi";
import { IdentifierRef, isUriRef, ObjRef } from "@gooddata/sdk-model";
import { TigerAuthenticatedCallGuard } from "../../../types";
import { convertFact, convertMeasure } from "../../../convertors/fromBackend/CatalogConverter";
import { TigerWorkspaceCatalog } from "./catalog";
import { loadAttributesAndDateDatasets } from "./datasetLoader";
import flatten from "lodash/flatten";
import flatMap from "lodash/flatMap";
import uniqBy from "lodash/uniqBy";
import sortBy from "lodash/sortBy";
import { MetadataUtilities, ValidateRelationsHeader } from "@gooddata/api-client-tiger";
import invariant from "ts-invariant";

export class TigerWorkspaceCatalogFactory implements IWorkspaceCatalogFactory {
    constructor(
        private readonly authCall: TigerAuthenticatedCallGuard,
        public readonly workspace: string,
        public readonly options: IWorkspaceCatalogFactoryOptions = {
            types: ["attribute", "measure", "fact", "dateDataset"],
            excludeTags: [],
            includeTags: [],
        },
    ) {}

    public withOptions = (options: Partial<IWorkspaceCatalogFactoryOptions>): IWorkspaceCatalogFactory => {
        const newOptions = {
            ...this.options,
            ...options,
        };
        return new TigerWorkspaceCatalogFactory(this.authCall, this.workspace, newOptions);
    };

    public forDataset = (dataset: ObjRef): IWorkspaceCatalogFactory => {
        return this.withOptions({
            dataset,
        });
    };

    public forTypes = (types: CatalogItemType[]): IWorkspaceCatalogFactory => {
        return this.withOptions({
            types,
        });
    };

    public includeTags = (tags: ObjRef[]): IWorkspaceCatalogFactory => {
        return this.withOptions({
            includeTags: tags,
        });
    };

    public excludeTags = (tags: ObjRef[]): IWorkspaceCatalogFactory => {
        return this.withOptions({
            excludeTags: tags,
        });
    };

    public load = async (): Promise<IWorkspaceCatalog> => {
        const promises: Promise<CatalogItem[]>[] = [];

        if (this.options.types.includes("measure")) {
            promises.push(this.loadMeasures());
        }
        if (this.options.types.includes("fact")) {
            promises.push(this.loadFacts());
        }

        const includeAttributes = this.options.types.includes("attribute");
        const includeDateDatasets = this.options.types.includes("dateDataset");
        if (includeAttributes || includeDateDatasets) {
            promises.push(this.loadAttributesAndDates(includeAttributes, includeDateDatasets));
        }

        const loadersResults = await Promise.all(promises);

        const catalogItems: CatalogItem[] = sortBy(flatten<CatalogItem>(loadersResults), (item) =>
            this.getCatalogItemSortingKey(item)?.toUpperCase(),
        );

        const groups = this.extractGroups(catalogItems);

        return new TigerWorkspaceCatalog(this.authCall, this.workspace, groups, catalogItems, this.options);
    };

    private getCatalogItemSortingKey = (item: CatalogItem) => {
        if (isCatalogAttribute(item)) {
            return item.attribute.title;
        }
        if (isCatalogFact(item)) {
            return item.fact.title;
        }
        if (isCatalogMeasure(item)) {
            return item.measure.title;
        }
        return undefined;
    };

    private tagsToIdentifiers = (tags: ObjRef[]): string[] => {
        return tags.map((ref) => {
            // Tags cannot be accessed by any separate endpoint, so it doesn't make sense to reference them by uri.
            // We will likely change the tag type signature from ObjRef to plain string in the future.
            invariant(!isUriRef(ref), "Tags cannot be referenced by uri!");
            return ref.identifier;
        });
    };

    private loadAttributesAndDates = async (
        loadAttributes: boolean,
        loadDateDataSets: boolean,
    ): Promise<CatalogItem[]> => {
        const tags = this.tagsToIdentifiers(this.options.includeTags);
        return this.authCall((client) =>
            loadAttributesAndDateDatasets(client, this.workspace, tags, loadAttributes, loadDateDataSets),
        );
    };

    private loadMeasures = async (): Promise<ICatalogMeasure[]> => {
        const tags = this.tagsToIdentifiers(this.options.includeTags);
        const measures = await this.authCall((client) => {
            return MetadataUtilities.getAllPagesOf(
                client,
                client.workspaceObjects.getAllEntitiesMetrics,
                {
                    workspaceId: this.workspace,
                },
                { query: { tags: tags.join(",") }, headers: ValidateRelationsHeader },
            )
                .then(MetadataUtilities.mergeEntitiesResults)
                .then(MetadataUtilities.filterValidEntities);
        });

        return measures.data.map(convertMeasure);
    };

    private loadFacts = async (): Promise<ICatalogFact[]> => {
        const tags = this.tagsToIdentifiers(this.options.includeTags);
        const facts = await this.authCall((client) => {
            return MetadataUtilities.getAllPagesOf(
                client,
                client.workspaceObjects.getAllEntitiesFacts,
                { workspaceId: this.workspace },
                { query: { tags: tags.join(",") } },
            ).then(MetadataUtilities.mergeEntitiesResults);
        });

        return facts.data.map(convertFact);
    };

    // Groups are collected from all catalog entities.
    // There is no separate endpoint for the tags anymore.
    private extractGroups(catalogItems: CatalogItem[]): ICatalogGroup[] {
        const groupableItems = catalogItems.filter((item) => item.type !== "dateDataset");
        const allTags = flatMap(groupableItems, (item): ICatalogGroup[] => {
            return (item as IGroupableCatalogItemBase).groups.map((tag) => ({
                title: (tag as IdentifierRef).identifier,
                tag: tag,
            }));
        });
        return uniqBy(allTags, (tag) => tag.title);
    }
}
