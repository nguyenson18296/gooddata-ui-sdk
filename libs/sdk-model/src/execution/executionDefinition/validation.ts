// (C) 2021 GoodData Corporation
import groupBy from "lodash/groupBy";
import toPairs from "lodash/toPairs";
import invariant from "ts-invariant";

import { attributeLocalId, isAttribute } from "../attribute";
import { measureLocalId } from "../measure";

import { IExecutionDefinition } from "./index";

/**
 * Validates the {@link IExecutionDefinition} instance and throws if it is invalid.
 * @param definition - the definition to validate
 * @internal
 */
export function defValidate(definition: IExecutionDefinition): void {
    const itemsWithLocalId = [...definition.attributes, ...definition.measures];

    const groups = groupBy(itemsWithLocalId, (item) =>
        isAttribute(item) ? attributeLocalId(item) : measureLocalId(item),
    );

    toPairs(groups).forEach(([localId, items]) => {
        invariant(
            items.length === 1,
            `There are ${items.length} items with the same localId '${localId}'. Please make sure the attributes and measures in the execution definition have unique localIds.`,
        );
    });
}
