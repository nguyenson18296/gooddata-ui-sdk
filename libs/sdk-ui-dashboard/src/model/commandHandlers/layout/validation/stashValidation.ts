// (C) 2021 GoodData Corporation

import {
    DashboardItemDefinition,
    ExtendedDashboardItem,
    isStashedDashboardItemsId,
    StashedDashboardItemsId,
} from "../../../types/layoutTypes";
import { LayoutStash } from "../../../state/layout/layoutState";

export type StashValidationResult = {
    existing: StashedDashboardItemsId[];
    missing: StashedDashboardItemsId[];
    resolved: ExtendedDashboardItem[];
};

export function validateAndResolveStashedItems(
    stash: LayoutStash,
    items: ReadonlyArray<DashboardItemDefinition>,
): StashValidationResult {
    const result: StashValidationResult = {
        missing: [],
        existing: [],
        resolved: [],
    };

    items.forEach((item) => {
        if (!isStashedDashboardItemsId(item)) {
            result.resolved.push(item);
            return;
        }

        if (stash[item] !== undefined) {
            result.existing.push(item);
            result.resolved.push(...stash[item]);
        } else {
            result.missing.push(item);
        }
    });

    return result;
}
