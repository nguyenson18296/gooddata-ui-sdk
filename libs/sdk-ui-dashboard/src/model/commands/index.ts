// (C) 2021 GoodData Corporation

import { LoadDashboard, RenameDashboard, ResetDashboard, SaveDashboard, SaveDashboardAs } from "./dashboard";
import {
    AddAttributeFilter,
    ChangeAttributeFilterSelection,
    ChangeDateFilterSelection,
    MoveAttributeFilter,
    RemoveAttributeFilters,
    SetAttributeFilterParent,
} from "./filters";
import {
    ChangeInsightWidgetFilterSettings,
    ChangeInsightWidgetHeader,
    ChangeInsightWidgetInsight,
    ChangeInsightWidgetVisProperties,
    ModifyDrillsForInsightWidget,
    RefreshInsightWidget,
    RemoveDrillsForInsightWidget,
} from "./insight";
import {
    ChangeKpiWidgetComparison,
    ChangeKpiWidgetFilterSettings,
    ChangeKpiWidgetHeader,
    ChangeKpiWidgetMeasure,
    RefreshKpiWidget,
} from "./kpi";
import {
    AddLayoutSection,
    AddSectionItems,
    ChangeLayoutSectionHeader,
    MoveLayoutSection,
    MoveSectionItem,
    RemoveLayoutSection,
    RemoveSectionItem,
    ReplaceSectionItem,
    UndoLayoutChanges,
} from "./layout";
import { CreateAlert, RemoveAlert, UpdateAlert } from "./alerts";
import { CreateScheduledEmail } from "./scheduledEmail";
import {
    Drill,
    DrillDown,
    DrillToAttributeUrl,
    DrillToCustomUrl,
    DrillToDashboard,
    DrillToInsight,
    DrillToLegacyDashboard,
} from "./drill";

export { DashboardCommandType, IDashboardCommand, CommandProcessingMeta } from "./base";
export {
    InitialLoadCorrelationId,
    LoadDashboard,
    loadDashboard,
    SaveDashboardAs,
    saveDashboardAs,
    SaveDashboard,
    saveDashboard,
    RenameDashboard,
    renameDashboard,
    ResetDashboard,
    resetDashboard,
} from "./dashboard";
export {
    ChangeDateFilterSelection,
    changeDateFilterSelection,
    clearDateFilterSelection,
    AddAttributeFilter,
    addAttributeFilter,
    MoveAttributeFilter,
    moveAttributeFilter,
    RemoveAttributeFilters,
    removeAttributeFilter,
    ChangeAttributeFilterSelection,
    AttributeFilterSelectionType,
    resetAttributeFilterSelection,
    changeAttributeFilterSelection,
    SetAttributeFilterParent,
    setAttributeFilterParent,
} from "./filters";
export {
    AddLayoutSection,
    addLayoutSection,
    MoveLayoutSection,
    moveLayoutSection,
    RemoveLayoutSection,
    removeLayoutSection,
    ChangeLayoutSectionHeader,
    changeLayoutSectionHeader,
    AddSectionItems,
    addSectionItem,
    ReplaceSectionItem,
    replaceSectionItem,
    MoveSectionItem,
    moveSectionItem,
    RemoveSectionItem,
    eagerRemoveSectionItem,
    UndoLayoutChanges,
    undoLayoutChanges,
    revertLastLayoutChange,
    DashboardLayoutCommands,
    UndoPointSelector,
} from "./layout";
export { CreateAlert, createAlert, RemoveAlert, removeAlert, UpdateAlert, updateAlert } from "./alerts";
export { CreateScheduledEmail } from "./scheduledEmail";
export {
    Drill,
    DrillDown,
    DrillToAttributeUrl,
    DrillToCustomUrl,
    DrillToDashboard,
    DrillToInsight,
    DrillToLegacyDashboard,
    drill,
    drillDown,
    drillToAttributeUrl,
    drillToCustomUrl,
    drillToDashboard,
    drillToInsight,
    drillToLegacyDashboard,
} from "./drill";
export {
    ChangeKpiWidgetHeader,
    changeKpiWidgetHeader,
    ChangeKpiWidgetMeasure,
    changeKpiWidgetMeasure,
    ChangeKpiWidgetFilterSettings,
    changeKpiWidgetFilterSettings,
    ChangeKpiWidgetComparison,
    changeKpiWidgetComparison,
    RefreshKpiWidget,
    refreshKpiWidget,
    KpiWidgetComparison,
} from "./kpi";

export {
    ChangeInsightWidgetHeader,
    changeInsightWidgetHeader,
    ChangeInsightWidgetFilterSettings,
    changeInsightWidgetFilterSettings,
    ChangeInsightWidgetVisProperties,
    changeInsightWidgetVisProperties,
    ChangeInsightWidgetInsight,
    changeInsightWidgetInsight,
    ModifyDrillsForInsightWidget,
    modifyDrillForInsightWidget,
    RemoveDrillsForInsightWidget,
    removeDrillForInsightWidget,
    RemoveDrillsSelector,
    RefreshInsightWidget,
    refreshInsightWidget,
} from "./insight";

/**
 * @internal
 */
export type DashboardCommands =
    | LoadDashboard
    | SaveDashboard
    | SaveDashboardAs
    | RenameDashboard
    | ResetDashboard
    | ChangeDateFilterSelection
    | AddAttributeFilter
    | RemoveAttributeFilters
    | MoveAttributeFilter
    | ChangeAttributeFilterSelection
    | SetAttributeFilterParent
    | AddLayoutSection
    | MoveLayoutSection
    | RemoveLayoutSection
    | ChangeLayoutSectionHeader
    | AddSectionItems
    | ReplaceSectionItem
    | MoveSectionItem
    | RemoveSectionItem
    | UndoLayoutChanges
    | ChangeKpiWidgetHeader
    | ChangeKpiWidgetMeasure
    | ChangeKpiWidgetFilterSettings
    | ChangeKpiWidgetComparison
    | RefreshKpiWidget
    | ChangeInsightWidgetHeader
    | ChangeInsightWidgetFilterSettings
    | ChangeInsightWidgetVisProperties
    | ChangeInsightWidgetInsight
    | ModifyDrillsForInsightWidget
    | RemoveDrillsForInsightWidget
    | RefreshInsightWidget
    | CreateAlert
    | UpdateAlert
    | RemoveAlert
    | CreateScheduledEmail
    | Drill
    | DrillDown
    | DrillToAttributeUrl
    | DrillToCustomUrl
    | DrillToDashboard
    | DrillToInsight
    | DrillToLegacyDashboard;
