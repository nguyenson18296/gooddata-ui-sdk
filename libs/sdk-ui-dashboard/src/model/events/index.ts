// (C) 2021 GoodData Corporation
import {
    DashboardLoaded,
    DashboardSaved,
    DateFilterValidationFailed,
    DashboardCopySaved,
    DashboardRenamed,
    DashboardWasReset,
} from "./dashboard";
import {
    DashboardCommandFailed,
    DashboardCommandRejected,
    DashboardQueryCompleted,
    DashboardQueryFailed,
    DashboardQueryRejected,
    DashboardQueryStarted,
} from "./general";
import {
    DashboardAttributeFilterAdded,
    DashboardAttributeFilterMoved,
    DashboardAttributeFilterParentChanged,
    DashboardAttributeFilterRemoved,
    DashboardAttributeFilterSelectionChanged,
    DashboardDateFilterSelectionChanged,
    DashboardFilterContextChanged,
} from "./filters";
import {
    DashboardLayoutChanged,
    DashboardLayoutSectionAdded,
    DashboardLayoutSectionHeaderChanged,
    DashboardLayoutSectionItemMoved,
    DashboardLayoutSectionItemRemoved,
    DashboardLayoutSectionItemReplaced,
    DashboardLayoutSectionItemsAdded,
    DashboardLayoutSectionMoved,
    DashboardLayoutSectionRemoved,
} from "./layout";
import {
    DashboardKpiWidgetChanged,
    DashboardKpiWidgetComparisonChanged,
    DashboardKpiWidgetFilterSettingsChanged,
    DashboardKpiWidgetHeaderChanged,
    DashboardKpiWidgetMeasureChanged,
} from "./kpi";
import {
    DashboardInsightWidgetChanged,
    DashboardInsightWidgetDrillsModified,
    DashboardInsightWidgetDrillsRemoved,
    DashboardInsightWidgetFilterSettingsChanged,
    DashboardInsightWidgetHeaderChanged,
    DashboardInsightWidgetInsightSwitched,
    DashboardInsightWidgetVisPropertiesChanged,
} from "./insight";
import { DashboardAlertCreated, DashboardAlertUpdated, DashboardAlertRemoved } from "./alerts";
import { DashboardScheduledEmailCreated } from "./scheduledEmail";
import { DashboardUserInteractionTriggered } from "./userInteraction";
import { Action } from "@reduxjs/toolkit";
import {
    DashboardRenderRequested,
    DashboardAsyncRenderRequested,
    DashboardAsyncRenderResolved,
    DashboardRenderResolved,
} from "./render";
import {
    DashboardDrillDownResolved,
    DashboardDrillToAttributeUrlResolved,
    DashboardDrillToCustomUrlResolved,
    DashboardDrillToDashboardResolved,
    DashboardDrillToInsightResolved,
    DashboardDrillToLegacyDashboardResolved,
    DashboardDrillResolved,
    DashboardDrillDownRequested,
    DashboardDrillToAttributeUrlRequested,
    DashboardDrillToCustomUrlRequested,
    DashboardDrillToDashboardRequested,
    DashboardDrillToInsightRequested,
    DashboardDrillToLegacyDashboardRequested,
    DashboardDrillRequested,
} from "./drill";

export { IDashboardEvent, DashboardEventType, isDashboardEvent } from "./base";
export {
    DateFilterValidationFailed,
    DateFilterValidationResult,
    DashboardLoaded,
    DashboardSaved,
    DashboardCopySaved,
    DashboardRenamed,
    DashboardWasReset,
} from "./dashboard";
export {
    DashboardCommandRejected,
    DashboardCommandFailed,
    ActionFailedErrorReason,
    isDashboardCommandFailed,
    DashboardQueryRejected,
    DashboardQueryFailed,
    DashboardQueryStarted,
    DashboardQueryCompleted,
    isDashboardQueryFailed,
} from "./general";

export {
    DashboardDateFilterSelectionChanged,
    DashboardFilterContextChanged,
    DashboardAttributeFilterParentChanged,
    DashboardAttributeFilterRemoved,
    DashboardAttributeFilterSelectionChanged,
    DashboardAttributeFilterMoved,
    DashboardAttributeFilterAdded,
} from "./filters";

export {
    DashboardLayoutSectionAdded,
    DashboardLayoutSectionMoved,
    DashboardLayoutSectionRemoved,
    DashboardLayoutSectionHeaderChanged,
    DashboardLayoutSectionItemsAdded,
    DashboardLayoutSectionItemReplaced,
    DashboardLayoutSectionItemMoved,
    DashboardLayoutSectionItemRemoved,
    DashboardLayoutChanged,
} from "./layout";

export {
    DashboardKpiWidgetHeaderChanged,
    DashboardKpiWidgetMeasureChanged,
    DashboardKpiWidgetFilterSettingsChanged,
    DashboardKpiWidgetComparisonChanged,
    DashboardKpiWidgetChanged,
} from "./kpi";

export {
    DashboardInsightWidgetHeaderChanged,
    DashboardInsightWidgetFilterSettingsChanged,
    DashboardInsightWidgetVisPropertiesChanged,
    DashboardInsightWidgetInsightSwitched,
    DashboardInsightWidgetDrillsModified,
    DashboardInsightWidgetDrillsRemoved,
    DashboardInsightWidgetChanged,
} from "./insight";
export { DashboardAlertCreated, DashboardAlertRemoved, DashboardAlertUpdated } from "./alerts";
export { DashboardScheduledEmailCreated } from "./scheduledEmail";
export {
    DashboardDrillDownRequested,
    DashboardDrillToAttributeUrlRequested,
    DashboardDrillToCustomUrlRequested,
    DashboardDrillToDashboardRequested,
    DashboardDrillToInsightRequested,
    DashboardDrillToLegacyDashboardRequested,
    DashboardDrillRequested,
    drillDownRequested,
    drillToAttributeUrlRequested,
    drillToCustomUrlRequested,
    drillToDashboardRequested,
    drillToInsightRequested,
    drillToLegacyDashboardRequested,
    drillRequested,
    DashboardDrillResolved,
    drillResolved,
    DashboardDrillDownResolved,
    DashboardDrillToAttributeUrlResolved,
    DashboardDrillToCustomUrlResolved,
    DashboardDrillToDashboardResolved,
    DashboardDrillToInsightResolved,
    DashboardDrillToLegacyDashboardResolved,
    drillDownResolved,
    drillToAttributeUrlResolved,
    drillToCustomUrlResolved,
    drillToDashboardResolved,
    drillToInsightResolved,
    drillToLegacyDashboardResolved,
} from "./drill";

export { DrillTargetsAdded, drillTargetsAdded } from "./drillTargets";

export { DashboardUserInteractionTriggered, userInteractionTriggered } from "./userInteraction";
export {
    DashboardRenderRequested,
    DashboardAsyncRenderRequested,
    DashboardAsyncRenderResolved,
    DashboardRenderResolved,
    asyncRenderRequested,
    asyncRenderResolved,
    renderRequested,
    renderResolved,
} from "./render";
/**
 * @alpha
 */
export type DashboardEvents =
    | DashboardLoaded
    | DateFilterValidationFailed
    | DashboardCommandFailed
    | DashboardCommandRejected
    | DashboardQueryFailed
    | DashboardQueryRejected
    | DashboardQueryStarted
    | DashboardQueryCompleted<any, any>
    | DashboardSaved
    | DashboardCopySaved
    | DashboardRenamed
    | DashboardWasReset
    | DashboardRenderRequested
    | DashboardAsyncRenderRequested
    | DashboardAsyncRenderResolved
    | DashboardRenderResolved
    | DashboardUserInteractionTriggered
    | DashboardDateFilterSelectionChanged
    | DashboardAttributeFilterAdded
    | DashboardAttributeFilterRemoved
    | DashboardAttributeFilterMoved
    | DashboardAttributeFilterSelectionChanged
    | DashboardAttributeFilterParentChanged
    | DashboardFilterContextChanged
    | DashboardLayoutSectionAdded
    | DashboardLayoutSectionMoved
    | DashboardLayoutSectionRemoved
    | DashboardLayoutSectionHeaderChanged
    | DashboardLayoutSectionItemsAdded
    | DashboardLayoutSectionItemReplaced
    | DashboardLayoutSectionItemMoved
    | DashboardLayoutSectionItemRemoved
    | DashboardLayoutChanged
    | DashboardKpiWidgetHeaderChanged
    | DashboardKpiWidgetMeasureChanged
    | DashboardKpiWidgetFilterSettingsChanged
    | DashboardKpiWidgetComparisonChanged
    | DashboardKpiWidgetChanged
    | DashboardInsightWidgetHeaderChanged
    | DashboardInsightWidgetFilterSettingsChanged
    | DashboardInsightWidgetVisPropertiesChanged
    | DashboardInsightWidgetInsightSwitched
    | DashboardInsightWidgetDrillsModified
    | DashboardInsightWidgetDrillsRemoved
    | DashboardInsightWidgetChanged
    | DashboardAlertCreated
    | DashboardAlertRemoved
    | DashboardAlertUpdated
    | DashboardScheduledEmailCreated
    | DashboardDrillDownResolved
    | DashboardDrillToAttributeUrlResolved
    | DashboardDrillToCustomUrlResolved
    | DashboardDrillToDashboardResolved
    | DashboardDrillToInsightResolved
    | DashboardDrillToLegacyDashboardResolved
    | DashboardDrillResolved
    | DashboardDrillDownRequested
    | DashboardDrillToAttributeUrlRequested
    | DashboardDrillToCustomUrlRequested
    | DashboardDrillToDashboardRequested
    | DashboardDrillToInsightRequested
    | DashboardDrillToLegacyDashboardRequested
    | DashboardDrillRequested;

/**
 * Creates DashboardEvent predicate that test whether the provided event matches it.
 *
 * @alpha
 * @param eventType - dashboard event type
 * @param pred - predicate to test
 * @returns boolean - matches?
 */
export function newDashboardEventPredicate<T extends DashboardEvents["type"]>(
    eventType: T,
    pred?: (event: DashboardEvents & { type: T }) => boolean,
) {
    return (event: Action): boolean => {
        if (event?.type !== eventType) {
            return false;
        }
        return pred?.(event as DashboardEvents & { type: T }) ?? true;
    };
}
