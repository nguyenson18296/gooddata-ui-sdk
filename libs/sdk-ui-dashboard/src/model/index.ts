// (C) 2021 GoodData Corporation

/*
 * The public API of the Dashboard model is exported from here.
 *
 * What is exported:
 *
 * -  hooks do dispatch commands / call selectors
 * -  all selectors & the typing of state with which they work
 * -  all events & their types. Note: event factories are not exported on purpose. outside code should not be
 *    creating events
 * -  all commands, their types & command factories
 */

export { useDashboardSelector, useDashboardDispatch, DashboardDispatch } from "./state/dashboardStore";
export { DashboardState } from "./state/types";
export { ObjRefMap, ObjRefMapConfig } from "./state/_infra/objRefMap";

export { selectDashboardLoading } from "./state/loading/loadingSelectors";
export { LoadingState } from "./state/loading/loadingState";
export { BackendCapabilitiesState } from "./state/backendCapabilities/backendCapabilitiesState";
export { selectBackendCapabilities } from "./state/backendCapabilities/backendCapabilitiesSelectors";
export { ConfigState } from "./state/config/configState";
export {
    selectConfig,
    selectLocale,
    selectSeparators,
    selectSettings,
    selectColorPalette,
    selectDateFilterConfig,
    selectObjectAvailabilityConfig,
    selectIsReadOnly,
    selectMapboxToken,
    selectDateFormat,
    selectEnableKPIDashboardSchedule,
    selectEnableKPIDashboardScheduleRecipients,
    selectEnableCompanyLogoInEmbeddedUI,
    selectIsEmbedded,
    selectPlatformEdition,
} from "./state/config/configSelectors";
export { PermissionsState } from "./state/permissions/permissionsState";
export {
    selectPermissions,
    selectCanListUsersInWorkspace,
    selectCanManageWorkspace,
} from "./state/permissions/permissionsSelectors";
export { FilterContextState } from "./state/filterContext/filterContextState";
export {
    selectFilterContext,
    selectFilterContextFilters,
    selectFilterContextDateFilter,
    selectFilterContextAttributeFilters,
} from "./state/filterContext/filterContextSelectors";

export { UndoEnhancedState, UndoEntry } from "./state/_infra/undoEnhancer";
export { LayoutState, LayoutStash } from "./state/layout/layoutState";
export {
    selectLayout,
    selectStash,
    selectBasicLayout,
    selectWidgetByRef,
} from "./state/layout/layoutSelectors";
export { DateFilterConfigState } from "./state/dateFilterConfig/dateFilterConfigState";
export {
    selectDateFilterConfigOverrides,
    selectEffectiveDateFilterConfig,
    selectEffectiveDateFilterTitle,
    selectEffectiveDateFilterMode,
    selectEffectiveDateFilterOptions,
    selectEffectiveDateFilterAvailableGranularities,
} from "./state/dateFilterConfig/dateFilterConfigSelectors";
export {
    selectInsights,
    selectInsightRefs,
    selectInsightsById,
    selectInsightByRef,
} from "./state/insights/insightsSelectors";
export { CatalogState } from "./state/catalog/catalogState";
export {
    selectAttributesWithDrillDown,
    selectCatalogAttributes,
    selectCatalogDateDatasets,
    selectCatalogFacts,
    selectCatalogMeasures,
    selectAllCatalogAttributesMap,
    selectAllCatalogDisplayFormsMap,
} from "./state/catalog/catalogSelectors";
export { AlertsState } from "./state/alerts/alertsState";
export { selectAlerts } from "./state/alerts/alertsSelectors";
export { UserState } from "./state/user/userState";
export { selectUser } from "./state/user/userSelectors";
export { DashboardMeta, DashboardMetaState } from "./state/meta/metaState";
export { selectDashboardRef, selectDashboardUriRef, selectDashboardTitle } from "./state/meta/metaSelectors";
export { selectListedDashboards } from "./state/listedDashboards/listedDashboardsSelectors";

export { selectDateDatasetsForInsight } from "./queryServices/queryInsightDateDatasets";
export { selectInsightAttributesMeta } from "./queryServices/queryInsightAttributesMeta";

export {
    DashboardContext,
    ObjectAvailabilityConfig,
    DashboardConfig,
    ResolvedDashboardConfig,
} from "./types/commonTypes";
export {
    ExtendedDashboardItem,
    ExtendedDashboardWidget,
    InsightPlaceholderWidget,
    KpiPlaceholderWidget,
    DashboardItemDefinition,
    StashedDashboardItemsId,
    ExtendedDashboardLayoutSection,
    RelativeIndex,
} from "./types/layoutTypes";
export { WidgetFilterSettings, WidgetHeader } from "./types/widgetTypes";

export { DateFilterConfigValidationResult } from "./_staging/dateFilterConfig/validation";
export { CatalogDateAttributeWithDataset } from "./_staging/catalog/dateAttributeWithDatasetMap";

export * from "./commands";
export * from "./events";
export * from "./queries";
export { DashboardEventHandler } from "./events/eventHandler";
export { createDrillToSameDashboardHandler } from "./eventHandlers/createDrillToSameDashboardHandler";
