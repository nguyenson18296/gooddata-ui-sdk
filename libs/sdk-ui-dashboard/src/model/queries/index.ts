// (C) 2021 GoodData Corporation

import { QueryDateDatasetsForInsight } from "./insights";

export { IDashboardQuery, DashboardQueryType } from "./base";
export { QueryDateDatasetsForInsight, DateDatasetsForInsight, queryDateDatasetsForInsight } from "./insights";

/**
 * @internal
 */
export type DashboardQueries = QueryDateDatasetsForInsight;
