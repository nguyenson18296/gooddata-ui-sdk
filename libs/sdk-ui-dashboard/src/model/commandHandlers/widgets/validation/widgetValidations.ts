// (C) 2021 GoodData Corporation

import { ObjRefMap } from "../../../../_staging/metadata/objRefMap";
import { IInsightWidget, isInsightWidget, IWidget } from "@gooddata/sdk-backend-spi";
import { ObjRef, serializeObjRef } from "@gooddata/sdk-model";
import { IDashboardCommand } from "../../../commands";
import { invalidArgumentsProvided } from "../../../events/general";
import { DashboardContext } from "../../../types/commonTypes";

type CommandWithRef = IDashboardCommand & {
    payload: {
        ref: ObjRef;
    };
};

/**
 * Given list of all dashboard widgets and a command that contains a ref, this function tests that the `ref` is
 * a reference to an existing dashboard widget and that the existing widget is an insight widget.
 *
 * If the validation succeeds, the located insight widget will be returned. Otherwise an error will fly. The error
 * will be an instance of DashboardCommandFailed event - it can be propagated through the command handler all the
 * way to the root command handler saga.
 *
 * @param ctx - dashboard context, this will be included in the DashboardCommandFailed event
 * @param widgets - map of widgets on the dashboard
 * @param cmd - any command that has 'ref' in its payload
 */
export function validateExistingInsightWidget(
    widgets: ObjRefMap<IWidget>,
    cmd: CommandWithRef,
    ctx: DashboardContext,
): IInsightWidget {
    const {
        correlationId,
        payload: { ref },
    } = cmd;
    const widget = widgets.get(ref);

    if (!widget) {
        throw invalidArgumentsProvided(
            ctx,
            `Cannot find insight widget with ref: ${serializeObjRef(ref)}.`,
            correlationId,
        );
    }

    if (!isInsightWidget(widget)) {
        throw invalidArgumentsProvided(
            ctx,
            `Widget with ref: ${serializeObjRef(ref)} exists but is not an insight widget.`,
            correlationId,
        );
    }

    return widget;
}
