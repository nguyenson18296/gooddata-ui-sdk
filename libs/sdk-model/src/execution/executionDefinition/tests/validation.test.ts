// (C) 2021 GoodData Corporation

import { IExecutionDefinition } from "../index";
import { ActivityType, Department, Won, WinRate } from "../../../../__mocks__/model";
import { modifyAttribute } from "../../attribute/factory";
import { modifyMeasure } from "../../measure/factory";
import { IAttribute } from "../../attribute";
import { IMeasure } from "../../measure";
import { emptyDef } from "../factory";
import { defValidate } from "../validation";

describe("defValidate", () => {
    const Workspace = "testWorkspace";

    it("should NOT throw for valid execution definition", () => {
        const def: IExecutionDefinition = {
            ...emptyDef(Workspace),
            attributes: [ActivityType, Department],
            measures: [WinRate, Won],
        };
        expect(() => defValidate(def)).not.toThrow();
    });

    const InvalidScenarios: Array<[string, IMeasure[], IAttribute[]]> = [
        [
            "two measures with the same localId",
            [modifyMeasure(Won, (m) => m.localId("foo")), modifyMeasure(WinRate, (m) => m.localId("foo"))],
            [],
        ],
        [
            "two attributes with the same localId",
            [],
            [
                modifyAttribute(ActivityType, (a) => a.localId("foo")),
                modifyAttribute(Department, (a) => a.localId("foo")),
            ],
        ],
        [
            "measure and an attribute with the same localId",
            [modifyMeasure(Won, (m) => m.localId("foo"))],
            [modifyAttribute(Department, (a) => a.localId("foo"))],
        ],
    ];

    it.each(InvalidScenarios)("should throw for execution definition with %s", (_, measures, attributes) => {
        const def: IExecutionDefinition = { ...emptyDef(Workspace), attributes, measures };
        expect(() => defValidate(def)).toThrowErrorMatchingSnapshot();
    });
});
