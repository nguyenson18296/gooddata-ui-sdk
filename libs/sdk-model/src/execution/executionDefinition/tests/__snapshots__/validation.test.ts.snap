// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`defValidate should throw for execution definition with measure and an attribute with the same localId 1`] = `"There are 2 items with the same localId 'foo'. Please make sure the attributes and measures in the execution definition have unique localIds."`;

exports[`defValidate should throw for execution definition with two attributes with the same localId 1`] = `"There are 2 items with the same localId 'foo'. Please make sure the attributes and measures in the execution definition have unique localIds."`;

exports[`defValidate should throw for execution definition with two measures with the same localId 1`] = `"There are 2 items with the same localId 'foo'. Please make sure the attributes and measures in the execution definition have unique localIds."`;
